### ISSUE TITLE
`renderComponent` options API proposal

### ISSUE THREAD

**Issue by robbiepitts** [glimmerjs/glimmer-application#90]
It seems like a good thing generally to change the `renderComponent` API slightly to accept an options object as the third argument. However, there is a detail I have feels about. This issue can serve as a place to discuss this. I would like to propose the following signature:

```ts
renderComponent(
  component: string,
  parent: Simple.Node,
  options: Dict<any> = {}
): RenderComponentResult
```

The only option currently supported would be `nextSibling` (`Simple.Node`). Basically, we would replace the third argument (which is currently an optional `nextSibling`) with an options object (`Dict<any>`).

The `Dict<any>` type is important to my proposal as I think it could be a mistake to constrain it to a particular interface. If we open it up to accept arbitrary options—combined with [#89](https://github.com/glimmerjs/glimmer-application/pull/89)—we give custom component manager authors a lot of power to experiment since they would now be able to access any arbitrary data they may need from the "outside".

https://github.com/glimmerjs/glimmer-application/pull/89 gives us access to the "root" object—produced by `renderComponent`—inside the `ComponentManager#create` method via the `DynamicScope` argument (the fourth argument). For example, if you did:

```ts
app.renderComponent('foo-bar', parentElement, { baz: 'qux' });
```

Then you could use this data in your custom component manager like:

```ts
class ComponentManager {
  // ...
  create(environment, definition, args, dynamicScope) {
    // ...
    let renderedRoot = dynamicScope.get('_renderedRoot').value();
    let baz = renderedRoot.options.baz;  // 'qux'
    // ...
  }
}
```

**Comment by tomdale** [comment]
I am opposed to an API that encourages component manager authors to tell people to put arbitrary options into the options bag because it dramatically limits our options (pardon the pun) in the future for introducing new API.

I want to understand the use cases here better, and in particular why it is valuable to have the ability to pass a configuration option only to the top-level component—it seems like you'd want the same capability for all components, which would lead me to believe whatever you might use this API for is better served by configuration on the component or template itself.

**Comment by robbiepitts** [comment]
@tomdale Great feedback, thank you.

The general thing I am trying to work out is how to provide state to the component manager. For example, for the web component use case I want to be able to set the `element` property on the Glimmer component instance to the custom element that the fragment is being rendered into (see https://github.com/glimmerjs/glimmer-web-component/pull/18 for more context on that). The reason accessing this element using the `didRenderLayout` hook is not a viable option is because if we want to use element properties to render the component, setting the `element` property on the component in `didRenderLayout` is too late. We need to be able to access the element _before_ rendering the layout.

In the web component scenario you are rendering each custom element with `renderComponent`, making each one top-level. And web-component-enabled Glimmer components have different semantics and thus should have a different component manager. For example, there is no such thing as `args` with custom elements, so the custom component manager I created does not set the `args` property on the component.

So this is my attempt at a generic API that supports providing data to the component manager to inject into the component.

**Comment by robbiepitts** [comment]
After giving this more thought throughout the day and receiving feedback from a number of people I am thinking this may not be such a great idea. I extrapolated my web component use case into "people may want to send arbitrary state into their component manager" and I really can't think of any other solid use cases that don't sound like `args`. So I'm beginning to think that exposing the bare minimum—the parent element—to the component manager may be the best option for now.

Closing, can reopen if anyone thinks there is still value in either of my above proposals.
